src/appengine/handlers/cron/grouper.py

```diff
@@ -80,13 +81,147 @@ def combine_testcases_into_group(testcase_1, testcase_2, testcase_map):
+      # Rule: Don't group the same testcase and use different combinations for
+      # comparisons.
+      if testcase_1_id <= testcase_2_id:
+        continue
+
+      # If both testcase have the same group id, then no work to do.
+      if testcase_1.group_id == testcase_2.group_id and testcase_1.group_id:
+        continue
+
+      # Rule: Check both testcase are under the same project.
+      if testcase_1.project_name != testcase_2.project_name:
+        continue
+
+      # Rule: Security bugs should never be grouped with functional bugs.
+      if testcase_1.security_flag != testcase_2.security_flag:
+        continue
+
+      # Rule: Follow different comparison rules when crash types is one of the
+      # ones that have unique crash state (custom ones specifically).
+      if (testcase_1.crash_type in data_types.CRASH_TYPES_WITH_UNIQUE_STATE or
+          testcase_2.crash_type in data_types.CRASH_TYPES_WITH_UNIQUE_STATE):
+
+        # For grouping, make sure that both crash type and state match.
+        if (testcase_1.crash_type != testcase_2.crash_type or
+            testcase_1.crash_state != testcase_2.crash_state):
+          continue
+
+      else:
+        # Rule: For functional bugs, compare for similar crash states.
+        if not testcase_1.security_flag:
+          crash_comparer = CrashComparer(testcase_1.crash_type,
+                                         testcase_2.crash_type)
+          if not crash_comparer.is_similar():
+            continue
+
+        # Rule: Check for crash state similarity.
+        crash_comparer = CrashComparer(testcase_1.crash_state,
+                                       testcase_2.crash_state)
+        if not crash_comparer.is_similar():
+          continue
+
+      combine_testcases_into_group(testcase_1, testcase_2, testcase_map)
+
+
+def _has_testcase_with_same_params(testcase, testcase_map):
+  """Return a bool whether there is another testcase with same params."""
+  for other_testcase_id in testcase_map:
+    # yapf: disable
+    if (testcase.project_name ==
+        testcase_map[other_testcase_id].project_name and
+        testcase.crash_state ==
+        testcase_map[other_testcase_id].crash_state and
+        testcase.crash_type ==
+        testcase_map[other_testcase_id].crash_type and
+        testcase.security_flag ==
+        testcase_map[other_testcase_id].security_flag and
+        testcase.one_time_crasher_flag ==
+        testcase_map[other_testcase_id].one_time_crasher_flag):
+      return True
+    # yapf: enable
+
+  return False
+
+
+def _shink_large_groups_if_needed(testcase_map):
+  """Shrinks groups that exceed a particular limit."""
+  group_id_with_testcases_map = {}
+
+  for testcase in six.itervalues(testcase_map):
+    if not testcase.group_id:
+      continue
+
+    if not testcase.group_id in group_id_with_testcases_map:
+      group_id_with_testcases_map[testcase.group_id] = []
+
+    # Prioritize testcase with issues over regular testcases.
```

Comment:
should we also prioritize reproducible over non-reproducible?

Reply:
Done!